docs/source/cuda/cuda_array_interface.rst

```diff
@@ -69,13 +69,249 @@ The following are optional entries:
+   details of the Producer and Consumer, they should be able to, at their
+   discretion, override some of the synchronization semantics of the interface
+   to reduce the synchronization overhead. Overriding synchronization semantics
+   implies that:
+
+   - The CAI design, and the design and implementation of the Producer and
+     Consumer do not specify or guarantee correctness with respect to data
+     races.
+   - Instead, the User is responsible for ensuring correctness with respect to
+     data races.
+4. All operations and exchange of data are orchestrated from a single host
+   thread. Use of multiple threads on the host is outside of the scope of this
+   version of the CAI. A future revision to the interface specification may
+   consider the use of multiple host threads once the current version and its
+   implementations have matured and become well-understood.
+
+
+Interface Requirements
+~~~~~~~~~~~~~~~~~~~~~~
+
+The ``stream`` entry enables Producers and Consumers to avoid hazards when
+exchanging data. Expected behaviour of the Consumer is as follows:
+
+* When ``stream`` is not present or is ``None``:
+
+  - No synchronization is required on the part of the Consumer.
+  - The Consumer may enqueue operations on the underlying data immediately on
+    any stream.
+
+* When ``stream`` is an integer, its value indicates the stream on which the
+  Producer may have in-progress operations on the data, and which the Consumer
+  is be expected to either:
+
+  - Synchronize on before accessing the data, or
+  - Enqueue operations in when accessing the data.
+
+  The Consumer can choose which mechanism to use, with the following
+  considerations:
+
+  - If the Consumer synchronizes on the provided stream prior to accessing the
+    data, then it must ensure that no computation can take place in the provided
+    stream until its operations in its own choice of stream have taken place.
+    This could be achieved by either:
+
+    - Placing a wait on an event in the provided stream that occurs once all
+      of the Consumer's operations on the data are completed, or
+    - Avoiding returning control to the user code until after its operations
+      on its own stream have completed.
+
+  - If the consumer chooses to only enqueue operations on the data in the
+    provided stream, then it may return control to the User code immediately
+    after enqueueing its work, as the work will all be serialized on the
+    exported array's stream. This is sufficient to ensure correctness even if
+    the User code were to induce the Producer to subsequently start enqueueing
+    more work on the same stream.
+
+* If the User has set the Consumer to ignore CAI synchronization semantics, the
+  Consumer may assume it can operate on the data immediately in any stream with
+  no further synchronization, even if the ``stream`` member has an integer
+  value.
+
+
+When exporting an array through the CAI, Producers must ensure that:
```

Comment:
There should probably be a requirement here that the attached `stream` is valid for the lifetime of the CAI view. Otherwise if you export a CAI and then destroy the attached stream, you get a user after free hazard. This would be no different than the requirement that the viewed device memory not be freed while the CAI view is active.

Reply:
I think this is addressed in the section "Lifetime management" in the same document? CAI doesn't handle object lifetime and it's user's responsibility to ensure this (https://github.com/numba/numba/pull/5323).

Reply:
I think people are probably used to worrying about the lifetime management of the data, but may not be used to thinking about stream objects also having a lifetime.

Reply:
Ah, it's probably a good point. Perhaps @gmarkall could add one sentence to "Lifetime management"?

Reply:
This is a good point - I will add something about lifetime management of streams, in at least one of this section and the Lifetime Management section.

Reply:
I've now added details on the lifetime management of streams in https://github.com/numba/numba/pull/5162/commits/9b25e18865400782ff888f377915308fb10f7576